docs/reference/limits.rst

```diff
@@ -0,0 +1,75 @@
+==============
+Limits
+==============
+
+This document provides hard and soft limits in ScyllaDB.
+
+Cluster and Node Limits
+-----------------------------
+
+.. list-table:: 
+   :widths: 50 50
+   :header-rows: 1
+
+   * - Component
+     - Limit 
+   * - Nodes per cluster
+     - Low hundreds
+   * - Node size
+     - 256 vcpu
+  
+See :ref:`Hardware Requirements <system-requirements-hardware>` for storage
+and memory requirements and limits.
+
+CQL Limits
+--------------
+
+.. list-table:: 
+   :widths: 50 50
+   :header-rows: 1
+
+   * - Component
+     - Limit
+   * - Keyspaces per cluster 
+     - Thousands (tested with 1000)
+   * - Tables per keyspace 
+     - Low thousands (tested with 5000)
+   * - Tables per cluster (including indexes)
+     - Thousands
+   * - Tables with CDC enabled
+     - No limit. CDC can be enabled for all tables in a cluster.
+   * - Materialized views and secondary indexes per table
+     - Low tens
+   * - Columns per table
+     - Hundreds
+   * - Columns per cluster
+     - Tens of thousands
+   * - Partition size
+     - Gigabytes
+   * - Rows per partition
+     - No limit
+   * - Row size
+     - Latency-related soft limit: 
+     
+       Hundreds of kilobytes (good latency) or megabytes (mediocre latency)
+   * - Key length
+     - 65533
+   * - Table / CF name length
+     - 48 characters
+   * - Keyspace name length
+     - 48 characters
+   * - Query parameters in a query
+     - 65535 (2^16-1)
+   * - Statements in a batch
+     - 65535 (2^16-1)
+   * - Fields in a tuple
+     - 32768 (2^15) (just a few fields, such as 2-10, are recommended)
+   * - Collection (List)
+     - ~2 billion (231)
```

Comment:
What is this `231`? Should it be `2^31`?
BTW where is this number coming from, 2 billion items in a collection will cause all sort of issues. It won't crash ScyllaDB but tail latencies will probably be terrible.
If these limits are the most ScyllaDB can shoulder without causing unavailability, then this is fine.

Reply:
By the way, ideally, we shouldn't be blindly documenting hearsay and copying stuff from Cassandra's documentation or grep'ing source code - we should write tests that confirm our beliefs about these limits. In https://github.com/scylladb/scylladb/pull/16779 I wrote tests for the key length issues and learned the differences between what we thought we had, and what we actually have. In https://github.com/scylladb/scylladb/issues/7745 I discovered that although we thought that the length of individual set items was no longer limited to 64KB, sizes above this size don't actually work in unprepared statements (they work in prepared statements). Who knows how many other problems we have like this...

Reply:
I vote for A. Lets move forward with the data we do know.

Reply:
Ok, I will queue.